### PR TITLE
User RBAC

### PR DESCRIPTION
**Pull request by bastjan**
## Summary

This PR adds a Kubernetes controller to the `control-api` binary which updates RBAC rules so users can edit themselves. This is done according to [Control API: User](https://kb.vshn.ch/appuio-cloud/references/architecture/control-api-user.html).

## Checklist

- [x] PR contains a single logical change (to build a better changelog).
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

### CODE REVIEW

File: controller.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+			"Enabling this will ensure there is only one active controller manager.")
+	probeAddr := cmd.Flags().String("health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	usernamePrefix := cmd.Flags().String("username-prefix", "", "Prefix prepended to username claims. Usually the same as \"--oidc-username-prefix\" of the Kubernetes API server")
+
+	cmd.Run = func(*cobra.Command, []string) {
+		scheme := runtime.NewScheme()
+		setupLog := ctrl.Log.WithName("setup")
+
+		utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+		utilruntime.Must(orgv1.AddToScheme(scheme))
```
Comment (glrf) [inline controller.go:50]: Technically you don't need `orgv1` right? (But doesn't hurt either so :shrug:)

File: controllers/user_controller.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	Scheme   *runtime.Scheme
+
+	// UserPrefix is the prefix applied to the user in the ClusterRoleBinding.subjects.name.
+	UserPrefix string
+}
+
+//+kubebuilder:rbac:groups=appuio.io,resources=users,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=appuio.io,resources=users/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterroles;clusterrolebindings,verbs=get;list;watch;create;update;patch
+
```
Comment (glrf) [inline controllers/user_controller.go:30]: You're missing permissions to record events

File: controllers/user_controller.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *UserReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&controlv1.User{}).
+		Complete(r)
+}
+
+func roleName(user controlv1.User) string {
```
Comment (glrf) [inline controllers/user_controller.go:116]: I know that's how it's defined in the requirements, but AFAIK users will (might) be able to choose their own name eventually and I already see someone calling themselves `namespace` and screwing with existing `namespace-owner` roles or similar. 

Do we care? If so maybe some prefixing could fix that?

**Review by glrf — APPROVED** [review]
LGTM :rocket:  Just some minor things. 

I'm not sure about the name collision, might be a non issue :shrug:

File: controller.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		Use: "controller",
+	}
+
+	zapfs := flag.NewFlagSet("zap", flag.ExitOnError)
+	opts := zap.Options{}
+	opts.BindFlags(flag.CommandLine)
+	cmd.Flags().AddGoFlagSet(zapfs)
+
+	metricsAddr := cmd.Flags().String("metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
+	enableLeaderElection := cmd.Flags().Bool("leader-elect", false,
```
Comment (simu) [inline controller.go:39]: Why don't we enable leader election by default?

**Review by simu — APPROVED** [review]
LGTM.

File: controller.go (hunk 1 above)
Comment (bastjan) [reply controller.go:50]: I don't need it. It was more of a precaution if we need it for the next feature. I don't mind removing or leaving it in. It's a thing you won't see in go tests though. You'll see it on first deploy 😅

File: controllers/user_controller.go (hunk 2 above)
Comment (bastjan) [reply controllers/user_controller.go:30]: Thanks 😅

File: controllers/user_controller.go (hunk 3 above)
Comment (bastjan) [reply controllers/user_controller.go:116]: Good point. I like the idea of prefixing the role.

File: controller.go (hunk 4 above)
Comment (bastjan) [reply controller.go:39]: It's the default from kubebuilder. I don't see a need for the controller to be HA but I also would not mind enabling it by default.

File: controller.go (hunk 1 above)
Comment (glrf) [reply controller.go:50]: Fair enough. IMO we can leave it

File: controller.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
+	probeAddr := cmd.Flags().String("health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	usernamePrefix := cmd.Flags().String("username-prefix", "", "Prefix prepended to username claims. Usually the same as \"--oidc-username-prefix\" of the Kubernetes API server")
+	rolePrefix := cmd.Flags().String("role-prefix", "control-api-user-", "Prefix prepended to generated cluster roles and bindings to prevent name collisions.")
```
Comment (bastjan) [inline controller.go:44]: @glrf Any ideas for a good prefix?

Comment (glrf) [reply controller.go:44]: I generally prefer something with `:` for roles. Like `control-api:user:`?

**Review by glrf — APPROVED** [review]

**Pull request merged**